Mekhanik evgenii/fix 1346 1

EvgeniiMekhanik opened the pull request.

EvgeniiMekhanik committed:
Add tests to check new 'client_mem' option

- Add tests to check new configuration option
- Add tests to check that request is blocked
  and connection is closed if `client_mem` is exceeded
- Add tests to check that connection is closed
  if `client_mem` is exceeded by ping.

EvgeniiMekhanik committed:
Enable tests to check flood prevention

Also remove some `tests.test_stress.*` tests
from tests_disabled_tcpseg/remote files because
all test_stress already disabled.

EvgeniiMekhanik committed:
Fix flucky test

If we cehck warning in dmesg we always shout use
dmesg.unlimited_rate_on_tempesta_node as a decorator.

EvgeniiMekhanik committed:
Do not use sniffer to check `server_forward_retries`

Now Tempesta FW do not use `tcp_push` to send data,
so we can't use sniffer and check count of `PUSH` packets,
more over it is not reliably, because this method rely on
the call of `tcp_push`. Check count of received requests
on server instead.

EvgeniiMekhanik committed:
Fix tests after http adjusting memory in Tempesta

Also add test to check `client_lru_size` option.

EvgeniiMekhanik committed:
Implement new fault injection tests.

Implement new fault injection tests to cover new
allocation code.

EvgeniiMekhanik committed:
Implement tests to check reconfiguring client mem option

EvgeniiMekhanik committed:
Fixes according last changes in Tempesta FW

- Do not check error response if client was blocked,
  because of memory limit exceeded during processing
  request.

RomanBelozerov commented on tests/tests_disabled.json:
```diff
@@ -189,10 +189,6 @@
             "reason": "The response time varies each time, and it is not possible to predict it exactly."
         },
-        {
-            "name": "tests.stress.test_stress.TestContinuationFlood",
```
Have you fixed this test?  I didn't see this test in the issue description (817).

EvgeniiMekhanik replied:
I add  client_mem 10000 20000; to the config of this test and show that with appropriate client_mem option this test pass and we can safely stay under continuation flood

RomanBelozerov commented on tests/client_mem/test_client_mem.py:
```diff
@@ -106,7 +107,16 @@ def test_request(self):
 
-        client.send_request(request, "403")
+        client.make_request(request)
+        if self.expect_response:
```
Let's use `expected_status_code` instead of `self.expect_response`. Just use `send_requst` with `expected_status_code=403` or `expected_status_code=None`. In this case you can remove this `if`.
```python3
client.make_request(request, expected_status_code)
```

EvgeniiMekhanik replied:
It was totally removed after last changes in Tempesta FW. We check client memory at the end of softirq short, when we have no active requests. We can only drop connection with TCP RST

RomanBelozerov commented on tests/client_mem/test_client_mem.py:
```diff
@@ -137,7 +138,11 @@ def test_response(self):
 
-        client.send_request(request, "403")
+        client.make_request(request)
+        if not run_config.TCP_SEGMENTATION:
```
Let's create two tests here: 
- first with TCP seg for local and remote setups and  we will expect a close connection only  (disable on TCP seg forever); 
- second without TCP seg for local and remote setups and we will expect a 403 response with a close connection (disable on TCP seg forever).

It is a separate cases. It should not has the same names, descriptions etc.

EvgeniiMekhanik replied:
For segmentation we drop request, not response. So we don't need this test at all.

RomanBelozerov commented on tests/multiple_clients/test_multiple_clients.py:
```diff
@@ -99,3 +99,14 @@ async def test_tcp_fin_timeout(self):
             )
+
+    async def test_client_memory(self):
+        config = self.get_tempesta().config.defconfig
```
You should add docstring here and assert. What does the test check?

EvgeniiMekhanik replied:
Remove this test. We have the same in TestSeveralClientsWithSmallLRUSize

RomanBelozerov commented on tests/client_mem/test_client_mem.py:
```diff
@@ -74,6 +75,39 @@ async def test_invalid(self, name, client_mem_config):
+    }
+
+    backends = [DEPROXY_SERVER]
+    expect_response = None
```
Let's use `expected_status_code="404"` or `expected_status_code=None` here and remove `if` in the test.

EvgeniiMekhanik replied:
It was totally removed after last changes in Tempesta FW. We check client memory at the end of softirq short, when we have no active requests. We can only drop connection with TCP RST

RomanBelozerov commented on tests/client_mem/test_client_mem.py:
```diff
@@ -0,0 +1,356 @@
+
+        for id_ in range(3):
+            client = self.get_client(f"deproxy-interface-{id_}")
+            client.start()
```
You can use `self.get_clients()` here or `ForEach` class from `herlpers/utils.py`
```python3
        for client in self.get_clients():
            client.start()
```
or
```python3
        clients = ForEach(self.get_clients())
        clients.start()
```
and then:
```python3
        request = client.create_request(method="GET", uri="/", headers=[])
        clients.make_requests(request * 10)
        await server.wait_for_requests(3 * 10, strict=True)
        await client.wait_for_response()
```

EvgeniiMekhanik replied:
Yes fixed

RomanBelozerov commented on tests/client_mem/test_client_mem.py:
```diff
@@ -0,0 +1,356 @@
+        for id_ in range(3):
+            client = self.get_client(f"deproxy-interface-{id_}")
+            for i in range(10):
+                request = client.create_request(method="GET", uri="/", headers=[])
```
Let's move the creating request before `for` cycle. It doesn't need to create a request every step.

EvgeniiMekhanik replied:
yes fixed

RomanBelozerov commented on tests/client_mem/test_client_mem.py:
```diff
@@ -0,0 +1,356 @@
+
+    @staticmethod
+    def make_resp(body):
+        return "HTTP/1.1 200 OK\r\n" "Content-Length: " + str(len(body)) + "\r\n\r\n" + body
```
```suggestion
        return f"HTTP/1.1 200 OK\r\nContent-Length: {len(body)}\r\n\r\n{body}"
```

RomanBelozerov commented on tests/client_mem/test_client_mem.py:
```diff
@@ -0,0 +1,356 @@
+
+    backends = [DEPROXY_SERVER]
+
+    def _ping(self, client):
```
Please add `send_ping` method to `DeproxyClientH2`.

EvgeniiMekhanik replied:
Yes fixed

RomanBelozerov commented on tests/client_mem/test_client_mem.py:
```diff
@@ -0,0 +1,356 @@
+"""Tests for client mem configuration."""
```
Please add docstring to all tests. In most cases, it is necessary to guess what conditions provide to this behavior.

EvgeniiMekhanik replied:
Yes fixed

RomanBelozerov commented on tests/client_mem/test_client_mem.py:
```diff
@@ -0,0 +1,356 @@
+            if i % 2 == 0:
+                config = base_config + "client_mem 10000 20000;\n"
+            else:
+                config = base_config
```
```suggestion
            config = base_config + "client_mem 10000 20000;\n" if i % 2 == 0 else base_config
```

RomanBelozerov commented on tests/client_mem/test_client_mem.py:
```diff
@@ -0,0 +1,356 @@
+    async def test_under_load(self):
+        await self.start_all_services(client=False)
+        client = self.get_client("gflood")
+        task = asyncio.create_task(self._do_reload())
```
Let's add `create_task` method to `tester.TempestaTest` which add all new tasks to a task list and cancel them in `cleanUp` methods after the test.
Why?
`self.stop = True` and `await task` don't call if this test fail in `await self.wait_while_busy(client)` or other steps and the task will work forever.

RomanBelozerov commented on tests/client_mem/test_client_mem.py:
```diff
@@ -0,0 +1,356 @@
+        },
+        {
+            "name": "Https",
+            "clients": [DEPROXY_CLIENT_SSL],
```
You created tests for Http, Https and Http2 in the PR. Are you sure we need these protocols? Are these not duplicates?

EvgeniiMekhanik replied:
I think yes, they are fast and we need to check that we block all types of clients

EvgeniiMekhanik committed:
Fix according review

- Add dockstrings for all new implemented tests
- Move `send_ping` function to DeproxyClientH2
- Use self.get_clients to iteration over all clients
- Remove unnecessary test
- Block some tests for TCP segmentation.
- Implement `create_task` in base TempestaTest class to
  be sure that all tasks will be finished successfully

RomanBelozerov commented on framework/deproxy/deproxy_client.py:
```diff
@@ -608,6 +608,11 @@ def make_request(
             self._valid_req_num += 1
 
+    def ping(self):
+        self.h2_connection.ping(opaque_data=b"\x00\x01\x02\x03\x04\x05\x06\x07")
```
Probably we should use data as function argument? For example:
```suggestion
    def send_ping(self, data: bytes =  b"\x00\x01\x02\x03\x04\x05\x06\x07") -> None:
        self.h2_connection.ping(opaque_data=data)
```
or `data: bytes = b""` if we can use empty data here

RomanBelozerov commented on framework/test_suite/tester.py:
```diff
@@ -160,6 +161,8 @@ class TempestaTest(WaitUntilAsserts, unittest.IsolatedAsyncioTestCase):
 
     clients: list[dict] = []
 
+    tasks: list[tuple(asyncio.Task, asyncio.event)] = []
```
Let's make the private variable. We should not have access to tasks list in out tests. And I think we should move the variable to `asyncSetUp` method. It is the variable for test, but not for test class etc. See `self.__oops`
```suggestion
    __tasks: list[asyncio.Task] = []
```

RomanBelozerov commented on framework/test_suite/tester.py:
```diff
@@ -375,6 +378,29 @@ def start_all_clients(self):
+            stop_event.set()
+            await task
+        self.tasks.clear()
+
```
You cannot use the `wrapper` here because the task may freeze in 387 line after first call. I think we can just create tasks and add it to task list and cancel after.
```suggestion
    def create_task(self, func) -> asyncio.Task:
        task = asyncio.create_task(func())
        self.tasks.append(task)
        return task

    async def __cleanup_tasks(self):
        for task in self.tasks:
            try:
                if not task.done():
                    task.cancel()
            except asyncio.CancelledError:
                ...
        self.tasks.clear()
```

RomanBelozerov commented on framework/test_suite/tester.py:
```diff
@@ -150,7 +150,8 @@ class TempestaTest(WaitUntilAsserts, unittest.IsolatedAsyncioTestCase):
     1) backends: [...]
     2) clients: [...]
-    3) several test functions.
+    3) tasks: [...]
```
Please back the description before changes

EvgeniiMekhanik committed:
Fix according review

- Use threading to create tasks.

RomanBelozerov commented on framework/test_suite/tester.py:
```diff
@@ -375,6 +376,26 @@ def start_all_clients(self):
+        def wrapper():
+            func(stop_event)
+
+        task = threading.Thread(target=wrapper)
```
optional

```suggestion
        task = threading.Thread(target=func, args=(stop_event, ))
```
